guppylang/prelude/_internal/checker.py

```diff
@@ -195,47 +206,107 @@ class NewArrayChecker(CustomCallChecker):
     """Function call checker for the `array.__new__` function."""
 
     def synthesize(self, args: list[ast.expr]) -> tuple[ast.expr, Type]:
-        if len(args) == 0:
-            raise GuppyTypeError(
-                "Cannot infer the array element type. Consider adding a type "
-                "annotation.",
-                self.node,
-            )
-        [fst, *rest] = args
-        fst, ty = ExprSynthesizer(self.ctx).synthesize(fst)
-        checker = ExprChecker(self.ctx)
-        for i in range(len(rest)):
-            rest[i], subst = checker.check(rest[i], ty)
-            assert len(subst) == 0, "Array element type is closed"
-        result_ty = array_type(ty, len(args))
-        call = GlobalCall(
-            def_id=self.func.id, args=[fst, *rest], type_args=result_ty.args
-        )
-        return with_loc(self.node, call), result_ty
+        match args:
+            case []:
+                raise GuppyTypeError(
+                    "Cannot infer the array element type. Consider adding a type "
+                    "annotation.",
+                    self.node,
+                )
+            # Either an array comprehension
+            case [DesugaredGeneratorExpr() as compr]:
+                return self.synthesize_array_comprehension(compr)
+            # Or a list of array elements
+            case [fst, *rest]:
+                fst, ty = ExprSynthesizer(self.ctx).synthesize(fst)
+                checker = ExprChecker(self.ctx)
+                for i in range(len(rest)):
+                    rest[i], subst = checker.check(rest[i], ty)
+                    assert len(subst) == 0, "Array element type is closed"
+                result_ty = array_type(ty, len(args))
+                call = GlobalCall(
+                    def_id=self.func.id, args=[fst, *rest], type_args=result_ty.args
+                )
+                return with_loc(self.node, call), result_ty
+            case args:
+                return assert_never(args)  # type: ignore[arg-type]
```

Comment:
Mypy isn't smart enough to infer that this code is unreachable, therefore `type: ignore[arg-type]`